Code/GraphMol/ChemReactions/SanitizeRxn.cpp

```diff
@@ -0,0 +1,349 @@
+      Atom *atom = (*it)->getAtomWithIdx(idx);
+      AtomInfo at(atom, templateIdx);
+      if (at.NeedsRLabel())
+        reactantAtomsToFix.push_back(at);
+    }
+  }
+  
+  templateIdx = 0;
+  for(MOL_SPTR_VECT::iterator it = rxn.beginProductTemplates();
+      it != rxn.endProductTemplates();
+      ++it, ++templateIdx) {
+    for(unsigned int idx=0;idx<(*it)->getNumAtoms(); ++idx) {
+      Atom *atom = (*it)->getAtomWithIdx(idx);
+      AtomInfo at(atom, templateIdx);
+      if (at.NeedsRLabel())
+        productAtomsToFix.push_back(at);
+    }
+  }
+
+  if (!reactantAtomsToFix.size() && !productAtomsToFix.size())
+    return;
+
+  if( reactantAtomsToFix.size() != productAtomsToFix.size() ) {
+    std::ostringstream str;
+    str << "Mismatched rlabels: " <<
+        reactantAtomsToFix.size() << " unmapped reactant rlabels," << 
+        productAtomsToFix.size() << " unmappped product rlabels" ;
+    throw RxnSanitizeException(str.str());
+  }
+
+
+  unsigned int max_rlabel = getMaxProp<unsigned int>(rxn,
+                                                     common_properties::_MolFileRLabel);
+  int max_atom_map = getMaxProp<int>(rxn,
+                                     common_properties::molAtomMapNumber);
+
+  BOOST_FOREACH(AtomInfo &rat, reactantAtomsToFix) {
+    bool found = false;
+    unsigned int bestGuess = rat.bestGuessRLabel();
+    if (!bestGuess) {
+      throw RxnSanitizeException(makeReactantErrorMessage(
+          "Could not deduce RLabel", rat));
+    }
+    
+    BOOST_FOREACH(AtomInfo &pat, productAtomsToFix) {
+      if (!pat.atom) continue;
+                         
+      if(rat.bestGuessRLabel() == pat.bestGuessRLabel()) {
+        // if the atomMaps don't match, this is bad, no atomMap is ok(==0)
+        if (rat.atomMap == pat.atomMap) {
+          found = true;
+          rat.setRLabel( max_rlabel + rat.bestGuessRLabel() );
+          pat.setRLabel( max_rlabel + pat.bestGuessRLabel() );
+          if (!rat.atomMap) { // set atom mapping as well
+            rat.setAtomMap(max_atom_map + rat.bestGuessRLabel());
+            pat.setAtomMap(max_atom_map + rat.bestGuessRLabel());
+          }
+          pat.atom = NULL;
+          break;
+        }
+      }
+    }
+    
+    if(!found) {
+      throw RxnSanitizeException(makeReactantErrorMessage(
+          "Could not find RLabel mapping", rat));
+    }
+  }
+  return;
+}
+
+// if we have query atoms without rlabels, make proper rlabels if possible
+//  ensure that every rlabel in the reactant has one in the product
+
+void fixAtomMaps(ChemicalReaction &rxn) {
+  int max_atom_map = getMaxProp<int>(
+      rxn,
+      common_properties::molAtomMapNumber);
+  std::map<unsigned int, int> potential_mappings;
+
+  unsigned int templateIdx = 0;
+
+  for(MOL_SPTR_VECT::iterator it = rxn.beginReactantTemplates();
+      it != rxn.endReactantTemplates();
+      ++it, ++templateIdx) {
+    for(unsigned int idx=0;idx<(*it)->getNumAtoms(); ++idx) {
```

Comment:
Why not just loop directly over atoms since you aren't using the index for anything?